rust/syntect-plugin/src/stackmap.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! StackMap is a simple nested map type (a trie) used to map `StackScope`s to
+//! u32s so they can be efficiently sent to xi-core.
+//!
+//! For discussion of this approach, see [this
+//! issue](https://github.com/google/xi-editor/issues/284).
+
+use std::collections::HashMap;
+
+use syntect::parsing::Scope;
+
+
+#[derive(Debug, Default)]
+struct Node {
+    value: Option<u32>,
+    children: HashMap<Scope, Node>,
+}
+
+#[derive(Debug, Default)]
+/// Nested lookup table for stacks of scopes.
+pub struct StackMap {
+    next_id: u32,
+    scopes: Node,
+}
+
+#[derive(Debug, PartialEq)]
+/// Result type for `StackMap` lookups. Used to communicate to the user
+/// whether or not a new identifier has been assigned, which will need to
+/// be communicated to the peer.
+pub enum LookupResult {
+    Existing(u32),
+    New(u32),
+}
+
+impl Node {
+    pub fn new(value: u32) -> Self {
+        Node { value: Some(value), children: HashMap::new() }
+    }
+
+    fn get_value(&mut self, stack: &[Scope], next_id: u32) -> LookupResult {
```

Comment:
As is I bet this lookup is slower than a `HashMap<Vec<Scope>,Node>` would be because of repeated pointer chasing costs. However, this tree structure is still good because if this is slow it's possible to use `ScopeStack::apply_with_hook` to find the basic pushes and pops of an operation, and use that to move a cursor around the tree with each operation, making the lookup time tiny and no longer proportional to stack depth.

The trie might also use less memory but I'm not certain.